compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp

```diff
@@ -644,8 +644,7 @@ struct ConvertMHLOToLinalgExtPass
         IREE::Flow::FlowDialect, mlir::cf::ControlFlowDialect,
         mlir::math::MathDialect, mlir::arith::ArithmeticDialect,
         tensor::TensorDialect, complex::ComplexDialect>();
-    target.addIllegalOp<mhlo::SortOp, mhlo::ScatterOp, mhlo::FftOp,
-                        mhlo::ReverseOp>();
+    target.addIllegalOp<mhlo::SortOp, mhlo::FftOp, mhlo::ReverseOp>();
```

Comment:
Wouldnt this mean that we dont handle scatter at all?

Reply:
As we don't mark it (or the dialect) as legal it will still get handled but just not flagged as illegal at end of pass. I saw all the same tests passing locally with this change so checking here too, if there is a specific one I should look at let me know.

Reply:
I thought that we should mark them to illegal dialect. But if it works, it's fine.

This PR does not really fix the issue... The main issue is that the frontend doesn't know how to lower the op. We'll have to lower it to Linalg/LinalgExt at some point. There is a `VerifyCompilerMHLOInputLegality` pass in the pipeline. It just moves the failure to the pass.

Reply:
All tests pass, but that makes me question the coverage on tests. We do have tests for mhlo.scatter.. So maybe its OK. My understanding was if the op is not marked illegal it isnt converted by the dialect conversion framework. Maybe that changed. Stamping (but my spidey sense are tingling on this one)

Reply:
I saw that it works with `--hlo-legalize-to-linalg`. The pattern (in `--hlo-legalize-to-linalg`) was added by me. It's added at very beginning stage. There were many issues about it (e.g., it's really fragile and limited, it introduces more loops which is expensive). We end up with introducing a LinalgExt dialect and implementing the linalg_ext.scatter op.

I'm not sure what's missing in the pattern, but we should fix the pattern in the pass.

Reply:
If we're in favor of moving signals to `VerifyCompilerMHLOInputLegality`, maybe we should delete this line.

Reply:
If we mark it legal no matches will be attempted, if we mark it illegal matches will be attempted and it will fail (even in partial), while if not marked either legal or illegal, the match will be attempted (verified with printf) but conversion won't fail. We should be able to delete all of these except if we want to never fall back to the generic/non-ext lowering to linalg for these ops and would rather this pass fails.

Re, moving signals, I'm not sure I follow, this doesn't seem to be a legality issue, one of coverage and optimality maybe.

Reply:
(the emitOpErrors in this pattern look problematic: notifyMatchFailure should be used for failed patterns, op errors only for op invariants unmet and that should be checked in verifier)

Reply:
Thanks for the explanation about matching. I wasn't aware of it. :)

I think we want to never fall back to generic/non-ext lowering for these ops, so let's keep the logic as it is. IMO, we should enhance the pattern to handle the case.

Reply:
> If we mark it legal no matches will be attempted, if we mark it illegal matches will be attempted and it will fail (even in partial), while if not marked either legal or illegal, the match will be attempted (verified with printf) but conversion won't fail. We should be able to delete all of these except if we want to never fall back to the generic/non-ext lowering to linalg for these ops and would rather this pass fails.

Is the fact that the match will be attempted when not marked legal or illegal a bug or a feature? (Asking more for my understanding)

> 
> Re, moving signals, I'm not sure I follow, this doesn't seem to be a legality issue, one of coverage and optimality maybe.

Reply:
Its a feature. Legal means skip in match, illegal means fail in check, if marked neither then it doesn't skip to attempt match nor fails if remaining.

I'm fine with skipping this, when we should just document here that this is the expectation (the use of partial conversion below made me unsure).

Reply:
Iirc, this used to be full conversion but it was changed to partial conversion with a following validation pass first to facilitate better debugging in failure (it is easier to debug a failed partial conversion), and then to have some follow-on conversions venir the final validation.

I don't have a strong opinion: just recounting history asI remember it.

Reply:
SG, now Hanhan's comment makes sense: so if remove these here and add to that verify pass instead then this pass would not fail verification and the verification pass would catch it instead. Is this the preference?

Reply:
It makes sense to me. (we should also mark these ops legal to avoid using patterns in mhlo repo?)

Reply:
If they are marked legal, then the conversion to `linalg_ext` wont trigger (since they are legal).

Reply:
I meant to mark them legal in `MHLOToLinalgOnTensors` pass, so we can make sure that we don't use upstream patterns.